pkg/provider/azure/keyvault/keyvault.go

```diff
@@ -0,0 +1,259 @@
+	anAzure := &Azure{
+		kube:      kube,
+		store:     store,
+		namespace: namespace,
+	}
+	anAzure.iAzure = anAzure
+	azClient, err := anAzure.newAzureClient(ctx)
+
+	if err != nil {
+		return nil, err
+	}
+
+	anAzure.baseClient = azClient
+	return anAzure, nil
+}
+
+// implement store.Client.GetSecret Interface.
+// retrieve a secret with the secret name defined in ref.Property in a specific keyvault with the name ref.Name.
+func (a *Azure) GetSecret(ctx context.Context, ref esv1alpha1.ExternalSecretDataRemoteRef) ([]byte, error) {
+	version := ""
+	var secretBundle []byte
+
+	if ref.Version != "" {
+		version = ref.Version
+	}
+	secretName := ref.Property
+	nameSplitted := strings.Split(secretName, "_")
+	getTags := false
+
+	if nameSplitted[len(nameSplitted)-1] == "TAG" {
+		secretName = nameSplitted[0]
+		getTags = true
+	}
+
+	secretMap, err := a.iAzure.getKeyVaultSecrets(ctx, ref.Key, version, secretName, getTags)
+	if err != nil {
+		return nil, err
+	}
+
+	secretBundle = secretMap[ref.Property]
+	return secretBundle, nil
+}
+
+// implement store.Client.GetSecretMap Interface.
+// retrieve ALL secrets in a specific keyvault with the name ref.Name.
+func (a *Azure) GetSecretMap(ctx context.Context, ref esv1alpha1.ExternalSecretDataRemoteRef) (map[string][]byte, error) {
+	secretMap, err := a.iAzure.getKeyVaultSecrets(ctx, ref.Key, ref.Version, "", true)
+	return secretMap, err
+}
+
+// getCertBundle returns the certificate bundle.
+func getCertBundleForPKCS(certificateRawVal string, certBundleOnly, certKeyOnly bool) (bundle string, err error) {
+	pfx, err := base64.StdEncoding.DecodeString(certificateRawVal)
+
+	if err != nil {
+		return bundle, err
+	}
+	blocks, _ := pkcs12.ToPEM(pfx, "")
+
+	for _, block := range blocks {
+		// skip the private key if looking for the cert only
+		if block.Type == "PRIVATE KEY" && certBundleOnly {
+			continue
+		}
+		// no headers
+		if block.Type == "PRIVATE KEY" {
+			pkey, err := x509.ParsePKCS1PrivateKey(block.Bytes)
+			if err != nil {
+				panic(err)
+			}
+			derStream := x509.MarshalPKCS1PrivateKey(pkey)
+			block = &pem.Block{
+				Type:  "RSA PRIVATE KEY",
+				Bytes: derStream,
+			}
+			if certKeyOnly {
+				bundle = string(pem.EncodeToMemory(block))
+				break
+			}
+		}
+
+		block.Headers = nil
+		bundle += string(pem.EncodeToMemory(block))
+	}
+	return bundle, nil
+}
+
+// consolidated method to retrieve secret value or secrets list based on whether or not a secret name is passed.
+// if the secret is of type PKCS then this is a cerificate that needs some decoding.
+func (a *Azure) getKeyVaultSecrets(ctx context.Context, vaultName, version, secretName string, withTags bool) (map[string][]byte, error) {
+	basicClient := a.baseClient
+	secretsMap := make(map[string][]byte)
+	certBundleOnly := false
+	certKeyOnly := false
+	secretNameinBE := secretName
+
+	if secretName != "" {
+		nameSplitted := strings.Split(secretName, "_")
+		if nameSplitted[len(nameSplitted)-1] == "CRT" {
```

Comment:
having a `CRT/KEY` extraction is certainly useful! IMO is this a generic feature and should be decoupled from a specific provider.

In the [crd spec](https://github.com/external-secrets/crd-spec/blob/main/Spec.md) we have the idea of a `template` which can use to interpolate and transform values. I think this would be a good place for it.

Reply:
I agree , it will be certainly useful .

But again comes the trade-off between making the CRD simple ,agnostic and generic for different backends , and between availing more features that can be available only for some stores and not for all (and this is a good example , Azure KV support saving certificates while other providers for example hashicorp vault dont AFAIK).

so we need to know how in which path shall we go .

Reply:
I've added a [comment ](https://github.com/external-secrets/crd-spec/pull/4#discussion_r576365504)to the related CRD Spec PR, if the provider returns relevant information on how to correctly decode the secret, then it makes sense to use this when parsing the secret. This would go under the `BestEffort / DefaultProvider` decode strategy.

I would propose leaving this implementation, and then we can take another pass after updating the CRD Spec and be sure the other decode strategy's are available and respected.